### ISSUE TITLE
Temperature check on possible memory leak in Holistic JS solution

### ISSUE THREAD

**Issue by codylieu** [google/mediapipe#1937]
Hello! I'm using the Holistic web solution in a WebRTC streaming application.

We've occasionally seen Holistic processing crash in a way that seems indicative of a memory leak:
![image](https://user-images.githubusercontent.com/6362795/115974931-47a7af00-a515-11eb-99d9-4e6b2194f4bd.png)

I'm currently unsure about whether the Holistic processing crash is the cause or a symptom of another issue. I've done some memory profiling, but haven't found a reliable way of reproducing it yet.

Since I don't have too much visibility into the Mediapipe JS internals, I was just hoping to get a temperature check on whether the Mediapipe team thinks:
1. This issue could be related to the Mediapipe internals / you've seen it before in other contexts
2. This is definitely not a Mediapipe issue and likely a bug in my application logic

Basically just trying to determine where to invest additional debugging / investigation efforts.

FWIW I also came across this thread https://bugs.chromium.org/p/chromium/issues/detail?id=1174675 which indicates there could be some memory leak issues in Chromium that can affect use cases like WebRTC, but the rate of leakage described there seems too slow compared to what I'm perceiving in my application.

Thanks in advance for your help! Sorry that I'm not able to provide any more details besides that single stack trace. Please let me know if you need any additional information and I'd be happy to circle back with it.

**Comment by mhays-google** [comment]
I see this.  Will check it out.

**Comment by mhays-google** [comment]
I have gone through our testing on Holistic and I see no memory leaks on the basic solution.  However, I may still be able to help.  WASM requires that any class created must subsequently be deleted to avoid memory leaks.

Can you check to see if it's possible that you are creating multiple copies of Solution without also calling `close()` on those solutions?  Remember that close() does return a Promise, so you will need to `await` or `.then()`.

**Comment by codylieu** [comment]
Thanks for the quick reply @mhays-google!

Your comment helped me find out that I was indeed instantiating multiple copies of Holistic in one of my flows. I'm not sure yet if this is also the root cause for the flow that originally led to my question though, but will continue digging and circle back. Just wanted to respond sooner than later to thank you for the useful advice!

Also just for my curiosity: not sure if this is expected since I don't have visibility into the `close()` implementation, but it didn't actually seem to have that big an impact on reclaiming memory. My problem was that I had a dangling reference that prevented garbage collection of the Holistic model between component renders. I tested both calling `close()` and just manually nulling the ref. The latter allowed the application to reclaim ~80 MB of memory (which seems to correlate the size of the instantiated Holistic object). The former didn't have a noticeable effect (though it's possible that I just didn't let it run long enough to gain a sizable memory footprint). Could you give me some more insight into what `close()` is actually doing underneath the hood? (the types / documentation you linked here https://github.com/google/mediapipe/issues/1408#issuecomment-810652766 are a bit vague)

**Comment by codylieu** [comment]
Hello again, @mhays-google

After making the above fix and doing some more profiling, it seems like there's still some seemingly unbounded memory allocation happening.

Here are some memory snapshots from my application:

1)
![image](https://user-images.githubusercontent.com/6362795/116836240-a060f280-ab7a-11eb-9a0d-1b7cba8f51d7.png)

2)
![image](https://user-images.githubusercontent.com/6362795/116836247-a8b92d80-ab7a-11eb-8043-192b62ae98de.png)

3)
![image](https://user-images.githubusercontent.com/6362795/116836261-b40c5900-ab7a-11eb-8dba-3caa9cc32a6a.png)

The culprit is always this `JsArrayBufferData` object that grows very large (I've seen this grow over 1.1 GB previously). This issue seems to be present in other Mediapipe models as well: https://github.com/google/mediapipe/issues/1956#issuecomment-830473628.

When I check the retainers that are keeping this `JsArrayBufferData` from getting garbage collected, the `createMediapipeSolutionsPackedAssets` and `createMediapipeSolutionsWasm` strings are the most notable to appear (the redacted blocks are just the name of my project just for privacy reasons)
![image](https://user-images.githubusercontent.com/6362795/116836578-e8344980-ab7b-11eb-8ea4-1231eeb79d5d.png)

Interestingly, I'm able to observe the same memory issue and retainers for this demo: https://mediapipe.dev/demo/holistic.
![image](https://user-images.githubusercontent.com/6362795/116836786-b8d20c80-ab7c-11eb-967d-4ddb3d2260e2.png)

Potentially notable similarities between my project and that demo: they both use the CDN method of import and therefore the `Holistic` object and `createMediapipeSolutionsPackedAssets` / `createMediapipeSolutionsWasm` functions are available globally in the Window.

Tangentially(?), I've since also tried the following to address this and neither have helped:
- Importing `Holistic` through the NPM package (https://www.npmjs.com/package/@mediapipe/holistic) so that `Holistic` isn't present in the Window (my `locateFile` still points to `https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.1.1617147418/${file}` so the `createMediapipe*` functions are still in the Window)
- Manually setting `createMediapipeSolutionsPackedAssets` and `createMediapipeSolutionsWasm` to `null` in my application to force garbage collection

Even more interestingly, the issue also seems to be present in https://mediapipe.dev/demo/holistic_remote/, albeit to a less extreme degree(?). Noteworthy: the implementation for this demo seems to differ from the previous based on what I can tell from the heap snapshot and the lack of `Holistic`, `createMediapipeSolutionsPackedAssets`, and `createMediapipeSolutionsWasm` in the Window
![image](https://user-images.githubusercontent.com/6362795/116837361-fdf73e00-ab7e-11eb-9465-2b7daddb6776.png)

Do you have any more clues / thoughts as to why this might be happening? Sorry for the block of text, just trying to give you as much information as I can about what I've tried so far :sweat_smile:

**Comment by mhays-google** [comment]
A few things:

`.close()` resolves to an equivalent call to `delete ptr` in C++.  Since these classes are allocated in C++ on the heap, they still need to be released.  Here, `JSArrayBufferData` represents your heap.  The fact that it continues to grow points specifically to new pointers being allocated in C++ without subsequently being deleted.

When you set a reference to `null` (or `undefined`) in JavaScript, it is able to garbage collect those objects which live exclusively in the JavaScript world.  However, if they are linked to any objects which are effectively proxies into some object living in C++, then you will no longer have the ability to tell C++ to release the resources.  It would be a bit like creating an object that adds a few items to a map, and then deleting that object before it has the opportunity to remove its contributions from said map.

Finally, it may be the case that heap memory grows but does not shrink.  If it does, we are not taking advantage of this (the same happens often in Java applications where they get the opportunity to grow, but never compact their own memory to allow shrinking it again).

Ok, with that out of the way:

`createMediapipeSolutionsPackedAssets` and `createMediapipeSolutionsWasm` should only ever be called once for a page.  The fact that they show up multiple times tells me that one of the following things have occurred:

1. (Most Likely) You are calling `send` multiple times without waiting for the returned Promise to resolve.  `send` will initialize the graph, which can potentially load the WASM.
2. You are manually calling `initialize` and not waiting for the Promise to resolve.
3. You are creating more than one Solution.

Can you check if any of these are the case?

Thank you!
Michael

**Comment by ben-xD** [comment]
Your comment has been useful to understand more about whats going on @mhays-google. I have a memory leak and have been trying to remove it for the past few days (the same issue with `JSArrayBufferData`), and I have checked that I am not doing any of the 3 issues you mentioned. 

> The fact that it continues to grow points specifically to new pointers being allocated in C++ without subsequently being deleted.
Isn't that done in C++, so shouldn't the mediapipe library be deleting these? Im not sure why deleting any variables in JS are going to affect this. `.close` isn't necessary because this memory leak happens _as the program runs_, and close is just called when you're done using the whole mediapipe solution. (i.e. call it right at the end)

> Finally, it may be the case that heap memory grows but does not shrink. If it does, we are not taking advantage of this 

Is this not the definition of a memory leak? Isn't this the problem all along? Could you clarify what you mean. If you the heap keeps growing because more arrays are created over time, I don't see how I can solve this in the javascript side.

PS: I also wish https://github.com/google/mediapipe/issues/1956 was not closed, because the issue was not addressed. @tyrmullen assumed it was one thing and closed it immediately. Though it is very related to this.

**Comment by tyrmullen** [comment]
Even if the memory being leaked is in C++ code, that doesn't necessarily mean that the MediaPipe library can fix this.  As mentioned by mhays-google, one key problem is that JS has no finalizers, which means it's unfortunately impossible for C++ code to know when garbage is collected.  The workaround for this is usually to force users to call some sort of .close() function as soon as resources are ok to be freed, on any JS object which is backed by C++ memory.

Also, the heap memory growing and never shrinking is not necessarily a memory leak, as long as there's a cap on the growth.  For example, if the C++ code starts out needing 10MB of heap space, but needs 20MB when a hand is visible, the heap size may leap to 20MB and then stay there, even if no hand is visible, so it technically only needs 10MB.  This can be expected behavior rather than a leak, since the next time you need 20MB, no more memory is allocated (and in fact, you usually don't want to be freeing and reallocating large chunks of memory on the fly, for performance reasons).

However, for your specific issue, I'm not actually convinced that the leak is even in the C++ memory...  I agree with your reading that there's only one C++ heap shown there, but 27MB doesn't seem too unreasonable pre-garbage-collection for a C++ heap size; is that particular value increasing steadily over time in your measurements?  Furthermore, when I run the smallest Holistic demo (without CodePen UI) [here](https://mediapipe.dev/demo/holistic), and then just look at Task Manager (right clicking on the headers there so I can enable "JavaScript memory", it looks like the "live" JavaScript memory never increases beyond a certain point (which would indicate a leak there).  However, the overall memory does appear to be increasing over time, even on this isolated MediaPipe demo.  So this makes me wonder where that other >100MB in your snapshots (which seems to be growing continuously) is coming from-- perhaps there are some DOM elements which are being kept alive in JS/TS code (or console.print logs) accidentally?  This will definitely require additional investigation.

On an unrelated note, I actually closed the other issue not as solved, but rather to deduplicate, precisely because it sounded like effectively the same problem as here, so it seemed more useful/efficient to have everything live on the same thread moving forwards (to "break the tie" I usually just choose the first issue reported, and close the others).

As for open-sourcing, this is a subject of much ongoing discussion and work.  More information can be found on issues [1408](https://github.com/google/mediapipe/issues/1408), [1741](https://github.com/google/mediapipe/issues/1741), and for some lower-level BUILD details [877](https://github.com/google/mediapipe/issues/877).

**Comment by ben-xD** [comment]
Thanks for your interesting reply 🙃

> doesn't necessarily mean that the MediaPipe library can fix this

I don't think we're expected to call `.close` on every frame, are we? I expected that once new Results for an image/ frame are returned from the library, that the old arrays are released in mediapipe/ **automagically**. I'm not sure of the distinction between the memory of mediapipe in JS and in C++ though.

- Even in my application (and the codepen), chrome's **Performance** tab's memory measurement is constant, it is the **memory** tab which shows it growing until 2GB and crashing, taking between 10 minutes to 20 minutes in my app. _It seems like you know why this is, is it because the performance tab only shows JS memory, and the memory dump shows more? I don't know 🤷🏻‍♂️_ Here is a picture of the Mediapipe codepen's memory growing? It does seem to grow in steps though, it was constant at 927MB for a few minutes, then suddenly 1028MB showed up.

![Screenshot 2021-05-05 at 10 12 42](https://user-images.githubusercontent.com/24711048/117119780-6ebe7600-ad8a-11eb-8451-02db7f3f7756.png)

> perhaps there are some DOM elements which are being kept alive in JS/TS code

Isn't the _other_ 100MB the _same_ 100MB we're talking about, they are ArrayBuffer's "retained" by `createMediapipeSolutionsPackedAssets`.

PS: Okay I understand why it was closed, but I replied there and I was afraid everything was going to waste in that issue.

**Comment by tyrmullen** [comment]
No, you shouldn't need to call .close() on every frame.  I was just explaining why the solution might not always be in C++ even if the apparent leak is in C++ memory; but aside from the Solution itself, I don't recall any of the intermediate results needing .close().  In general, anything which is a native JS data type (or a structure/dict/array of those) will not require a .close() call-- it would only be for something our API returns which keeps its functionality in C++ (so interacting with it would generally be through function calls-- like the Solution).

As for memory layout, the memory in C++ is basically a special siloed-off ArrayBuffer.  JS can peek into this and even manipulate it, but to the C++ code, this is the only memory it knows.  And I suspect you're right about the memory profiling differences-- Performance shows JS memory, while Memory dump shows everything.  In fact, if you check out Task Manager (with "JavaScript memory" as mentioned above), then I'd expect the "memory used" readings would grow and shrink with Memory dump values (but will likely be larger, since this covers stuff even the Memory dumps don't), while the "JavaScript memory" readings would correspond with Performance values.  And FWIW it looks like C++ memory might not count as "JavaScript memory", although I'm not quite sure where the distinction is drawn.

From your screenshot above, I see two different large chunks of memory, one which is ~27.5MB (@49035), and one which is positively massive (almost 1GB! (@419095)).  The expanded trace shows that the wasm Module (@87191) is causing ~27.7MB to be retained, 27.5MB of which I'm assuming would be from that smaller chunk of memory (@49035).  But I believe the wasm Module retains the entire native C++ heap, so it seems the C++ memory isn't growing beyond 27-ish MB.  So the question to ask now is "where is that ~1GB coming from?".

When I repeat your experiment with the various solutions, I only see a memory increase over time for "Holistic" and "FaceMesh", which makes me think this is facemesh-specific (so I'll look into FaceMesh alone next, to cut out some of the extra noise).  Curiously, though, my memory dump details look a little different than yours.  For me, the increase appears to clearly occur in the wasm Module's C++ memory area, indicating a leak there.

TL;DR: I'm not sure why we're getting different readouts, but at least for me, I seem to see a C++ memory leak somehow related to FaceMesh-- we don't do anything special for this module that I can think of, though, so I'm not yet sure what could cause this.

**Comment by ben-xD** [comment]
That's interesting, I haven't done a lot of WASM stuff but I'd be interested to help out.

> So the question to ask now is "where is that ~1GB coming from?".

I was trying to investigate that, but the codepen is broken, have the library/ model files have been removed?
```
Failed to load resource: the server responded with a status of 403 () https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3/holistic_solution_packed_assets.data
```

**Comment by codylieu** [comment]
Thanks @mhays-google! No need to eat crow haha, I'm just thankful that you and your team have made this model available at all!

**Comment by ben-xD** [comment]
> I'll have the memory leak fix up later today.

Was the fix included in yesterday's release `v0.3.1620694839`?, sadly I still observe a memory leak in my app. Unfortunately the [codepen](https://codepen.io/mediapipe/pen/LYRRYEw) is still broken, perhaps thats the old codepen and you created a new one?

I'm also curious why there are so many digits in the version string 👀

**Comment by afogel** [comment]
@mhays-google Looks like the holistic JS solution is still broken (version 0.3.1620694839):
here's the full stacktrace I'm getting:
```
holistic_solution_wasm_bin.js:9 Uncaught (in promise) RuntimeError: abort(undefined) at Error
    at jsStackTrace (https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.js:9:64184)
    at stackTrace (https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.js:9:64436)
    at abort (https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.js:9:37792)
    at _abort (https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.js:9:176986)
    at https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.wasm:wasm-function[9474]:0x508429
    at https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.wasm:wasm-function[9469]:0x5083c0
    at https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.wasm:wasm-function[9465]:0x5081ce
    at https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.wasm:wasm-function[9462]:0x507d6f
    at https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.wasm:wasm-function[9477]:0x508482
    at https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.wasm:wasm-function[259]:0xd815
    at abort (https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.js:9:37823)
    at _abort (https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.js:9:176986)
    at https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.wasm:wasm-function[9474]:0x508429
    at https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.wasm:wasm-function[9469]:0x5083c0
    at https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.wasm:wasm-function[9465]:0x5081ce
    at https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.wasm:wasm-function[9462]:0x507d6f
    at https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.wasm:wasm-function[9477]:0x508482
    at https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.wasm:wasm-function[259]:0xd815
    at https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.wasm:wasm-function[197]:0xb426
    at SolutionWasm$send [as send] (eval at new_ (https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.3.1620694839/holistic_solution_wasm_bin.js:9:157525), <anonymous>:9:1)
abort @ holistic_solution_wasm_bin.js:9
_abort @ holistic_solution_wasm_bin.js:9
(anonymous) @ holistic_solution_wasm_bin.wasm:0x508429
(anonymous) @ holistic_solution_wasm_bin.wasm:0x5083c0
(anonymous) @ holistic_solution_wasm_bin.wasm:0x5081ce
(anonymous) @ holistic_solution_wasm_bin.wasm:0x507d6f
(anonymous) @ holistic_solution_wasm_bin.wasm:0x508482
(anonymous) @ holistic_solution_wasm_bin.wasm:0xd815
(anonymous) @ holistic_solution_wasm_bin.wasm:0xb426
SolutionWasm$send @ VM2876:9
(anonymous) @ holistic.js:39
R @ holistic.js:10
next @ holistic.js:11
k @ holistic.js:22
Promise.then (async)
r @ camera_utils.js:7
(anonymous) @ camera_utils.js:7
requestAnimationFrame (async)
q @ camera_utils.js:7
a.video.onloadedmetadata @ camera_utils.js:7
```

**Comment by sgowroji** [comment]
Hi @afogel, Can you raise an another issue with complete details in it. 
@codylieu Thank you for the update we are closing this issue now.

**Comment by ben-xD** [comment]
@sgowroji why did you close this? This is an open issue.The fix was not released or confirmed by @mhays-google.

**Comment by afogel** [comment]
@sgowroji thanks for the fast response -- I added the notes here b/c @ben-xD had mentioned that the version referenced earlier in the thread meant to address the memory leak also seemed to break the holistic js model altogether. Given that it's been about 2 weeks since the release and the current release is broken, I figured I'd add the stack trace information if that might help resolve some of the issues.

**Comment by codylieu** [comment]
@sgowroji Just to clarify: we are still facing this issue. I was just thanking your team for all the work you've put into this

**Comment by codylieu** [comment]
Sorry, accidentally closed the issue. Reopened it. Just to clarify: we are still facing this issue @sgowroji. My comment was just thanking you and your team for all the work you've put into this so far!

**Comment by codylieu** [comment]
@mhays-google @sgowroji Any updates on this?

**Comment by tyrmullen** [comment]
The fundamental issue here is that the underlying memory management system is quite brittle and makes debugging leaks rather tricky. 
 If one "delete()" call is forgotten (and many are required for the various nested structures involved-- with no easy way to determine which are potentially missing), then the C++ memory is leaked opaquely, only visible when the leak is large enough or persists long enough.

This is definitely something we'd like to fix (and we have several ideas for doing so), but for most Solution APIs this isn't as bad a problem, since the data being brought from C++ back to JS is so very small (I don't think this issue is face_mesh- or holistic-specific; instead I believe that the face mesh data just happens to be the only data large enough to produce a noticeable memory leak within a reasonable amount of time).

That said, because face mesh data is so large, this results in MB/s being leaked, which is quite serious.  So while a memory management system restructuring will likely not occur for a while yet, I will attempt to track down and submit a patch in the next day or two to at least solve this particularly problematic case of face mesh data leakage when transferring from C++ to JS.

**Comment by tyrmullen** [comment]
I've submitted a patch-- the next JS Solutions update to "face_mesh" and "holistic" should no longer have this issue.  Please confirm or let me know if that's not the case when the next release occurs.

**Comment by sgowroji** [comment]
@tyrmullen Thank you so much for the update, Hi @afogel & @codylieu, Please test the above in the next release and reach us if you still face the similar issue. Thank you!

**Comment by afogel** [comment]
@sgowroji I just checked the npm repository, but haven't seen another release yet (The latest release I see is v. 0.3.1620694839, published a month ago)? Do you have a sense of when it will be released? Thanks!

**Comment by codylieu** [comment]
@sgowroji @tyrmullen Same question as @afogel. Still not seeing any new releases on: https://www.npmjs.com/package/@mediapipe/holistic. Any idea when a version with @tyrmullen's fix might be released?

**Comment by chuoling** [comment]
It should be there now in version 0.4.1625245411. Sorry for the delay and confusion.

**Comment by afogel** [comment]
@chuoling @sgowroji I'm not able to check to see whether the fix is working because I'm having the same issue as before on version 0.3.1620694839, though the error appears to occur in the `simd_wasm_bin.js` file rather than the `wasm_bin.js`.
I tried it on both Chrome and FireFox and it crashed. Can we reopen until this is resolved?
Thanks!
```
Uncaught (in promise) RuntimeError: abort(undefined) at jsStackTrace@https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.4.1625245411/holistic_solution_simd_wasm_bin.js:9:66465
stackTrace@https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.4.1625245411/holistic_solution_simd_wasm_bin.js:9:66717
abort@https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.4.1625245411/holistic_solution_simd_wasm_bin.js:9:39700
_abort@https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.4.1625245411/holistic_solution_simd_wasm_bin.js:9:178721
@https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.4.1625245411/holistic_solution_simd_wasm_bin.wasm:wasm-function[7398]:0x50a52e
@https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.4.1625245411/holistic_solution_simd_wasm_bin.wasm:wasm-function[7390]:0x50a243
@https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.4.1625245411/holistic_solution_simd_wasm_bin.wasm:wasm-function[7401]:0x50a5a3
@https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.4.1625245411/holistic_solution_simd_wasm_bin.wasm:wasm-function[253]:0x1eabb
@https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.4.1625245411/holistic_solution_simd_wasm_bin.wasm:wasm-function[191]:0x1657d
SolutionWasm$send@https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.4.1625245411/holistic_solution_simd_wasm_bin.js line 9 > Function:9:8
d/<@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:94005:401
W@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:93971:39
la/this.next@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:93972:89
h@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:93981:318
promise callback*f@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:93981:457
h@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:93981:315
promise callback*f@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:93981:457
Z/<@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:93981:468
Z@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:93981:261
parcelRequire<["../node_modules/@mediapipe/holistic/holistic.js"]</</A.send@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:94002:29
b/<@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:94017:399
W@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:93971:39
la/this.next@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:93972:89
Z/<@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:93981:491
Z@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:93981:261
parcelRequire<["../node_modules/@mediapipe/holistic/holistic.js"]</</A.send@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:94017:330
onFrame@file:///Users/repo/thv.js/dist/js/main.js:49:30
r@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:94081:858
q/<@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:94081:660
FrameRequestCallback*q@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:94081:626
t/a.video.onloadedmetadata@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:94081:754
EventHandlerNonNull*t@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:94081:702
parcelRequire<["../node_modules/@mediapipe/camera_utils/camera_utils.js"]</</p.prototype.start/<@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:94082:297
promise callback*parcelRequire<["../node_modules/@mediapipe/camera_utils/camera_utils.js"]</</p.prototype.start@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:94082:279
@file:///Users/repo/thv.js/dist/js/main.js:55:9
EventListener.handleEvent*$_ready@file:///Users/repo/thv.js/dist/engine/core/monogatari.js:18314:10
@file:///Users/repo/thv.js/dist/js/main.js:31:9
```

**Comment by tyrmullen** [comment]
@afogel I think we should open a separate issue for the error you're running into, since it seems unrelated to the memory leak, and we have other ways to verify the leak-fix (so it should not be blocking validation).  For example, the facemesh demo [mediapipe.dev/demo/face_mesh](http://mediapipe.dev/demo/face_mesh) should no longer have this severe memory leak.

**Comment by afogel** [comment]
done! #2267

**Comment by crecord** [comment]
I know this issue is closed, but I am struggling with memory leaks as described above.

I am using the [face mesh demo ](https://codepen.io/mediapipe/pen/KKgVaPJ) run on local server (no code pen). I changed nothing about the code. 

My versions are as follows: 
```
	camera_utils@0.3
	control_utils@0.6
	drawing_utils@0.3
	face_mesh@0.4
```

	Chrome Version 95.0.4638.69 (Official Build) (arm64)

I observed significant increases of JS heap size over time. I sampled every minute for 10 minutes and observed these increases.

![Screen Shot 2021-11-15 at 4 15 50 PM](https://user-images.githubusercontent.com/3160465/141856971-c832b54a-3a79-4cee-a455-0c6257326572.png)

`31.4,50.3,76.1,88.7,104,121,143,168,168,199`

I also observed this increase monitoring heap size over the course of 40 seconds.

![Screen Shot 2021-11-15 at 4 21 12 PM](https://user-images.githubusercontent.com/3160465/141857026-8f9630db-6a90-439e-8194-04c0d9d83e93.png)

Any suggestions on how to resolve this?

**Comment by vqc** [comment]
@tyrmullen the memory leak originally described by @codylieu and now by @crecord seems to exist again as well in pose@0.5.1635988162 cc: @mhays-google

**Comment by crecord** [comment]
I compare some different approaches [here](https://github.com/tensorflow/tfjs/issues/5854). It seems that using this [demo](https://www.notion.so/antimodularresearch/Banderoles-Dev-Notes-ddb6f2b1a8544d2391eb5f93939fcda9#c4c338e57d534191b212f8fc1e3df537) with NPM + the dev dependencies is the most promising for memory. Maybe a similar environment could help for using the facemesh mediapipe example.

**Comment by mhays-google** [comment]
Hmm... I've run code.mediapipe.dev/codepen/face_mesh in full screen mode
and in the editor.  I let it go for 10 minutes.  Not only was it not
leaking, but it was also hovering around 10 megabytes.

We definitely had leaks in the past (months ago), but we've addressed
this.  Is it possible that you have a bad version downloaded?
Specifically, in your constructor, are you maybe feeding a bad prefix to
`locateFile`?

**Comment by tyrmullen** [comment]
Unfortunately, I can verify that both of those demo scenarios leak on CodePen as well (current facemesh CodePen with no changes, and pose with version bumped to @0.5.1635988162).

Profiling makes it seem that there may even be two distinct leaks: one in C++/Wasm memory (`System/JsArrayBufferData`, which exhibits sporadic "leaps" in size) like before, and a much smaller one which appears to be in pure JS memory (`(array)`, which slowly continuously grows over time).  I believe the latter is related to rendering code, since if the tab is backgrounded (and thus should not be streaming in video frames), then it stops growing (while the other leak does not).

@mhays-google 10MB sounds very low for the facemesh demo (even when first starting up); maybe double-check to see if your memory profiler was accidentally set to profile the CodePen JVM and not the MP demo (the demo JVM should be the larger of the two)?  Also, you might not see most of the leaks unless you actually take heap snapshots, since the live JS heap report just on the memory profiler tab seems to ignore things like the C++/Wasm memory for some reason.

**Comment by vqc** [comment]
@tyrmullen and @mhays-google any update on this?
cc: @crecord

**Comment by shubham-dawande** [comment]
@tyrmullen @mhays-google 
Can we please get an update on the fix? We are experiencing crash issue on mobile devices due to this leak.
Facemesh version used: `0.4.1633559619`

**Comment by vqc** [comment]
@tyrmullen and @mhays-google any update on this?

thanks!
Vince

**Comment by tyrmullen** [comment]
I haven't yet had time to try to track this down, but I've at least isolated the regression.  Version 0.4.1625245753 had my initial large C++ leak fix, and no C++ memory appears to be leaked up through version 0.4.1628005669.  However, versions 0.4.1629159166 and now 0.4.1633559619 (most recent version as of this posting) both have a large C++ memory leak again.  For now, unless you need the new features, I'd recommend using one of those older versions as a workaround.

**Comment by shirlyyang-tcc** [comment]
> 

Is there any update? We met the same problem of momory leaking with facemesh in our web application, yet we do need the iris result so we can't work around with 0.4.1628005669. Is it possible to fix the problem?
